app/controllers/nodes_controller.rb

```diff
@@ -18,12 +18,31 @@ def show
     @minion = Minion.find(params[:id])
   end
 
+  def assign_roles
+    Minion.assign_roles(roles: { master: [params[:hostname]] })
+
+    respond_to do |format|
+      format.html { redirect_to nodes_path }
+      format.json { head :ok }
+    end
+  rescue Minion::CouldNotAssignRole,
```

Comment:
You shouldn't be using exceptions for code flow. There are a couple of issues here:

1. If `assign_roles` can raise an error and it performs an bang action, then it should be named with a bang `assign_roles!`. Otherwise, as a reader I have no way of knowing this (it's kind of a Rails convention 😉).
2. In my head, exceptions should really occur only in "exceptional" cases (well, I'm against exceptions in general, but that's a different story 😁). Not being able to assign a role or that the user wanted to change a non-existing minion is not that exceptional:
    - If the minion never existed, then this should be a 404. You could handle that by adding to this class (or the application_controller): `rescue_from Minion::NonExistingMinion, with: :not_found`. This way, you don't have to handle this explicitely, Rails and the frontend will already guess what a 404 stands for, and other methods that might face this situation, can act in the same way.
    - I'd actually expect the assign_roles method to return something meaningful (e.g. it can return whether it was successful or not, like Rails' `save` method). In this case, a good ol' if/else statement would fix this (note also that there's some shared code between these two cases: in html the only difference is in the flashy message).